Sources/_CryptoExtras/Key Derivation/PBKDF2/PBKDF2.swift

```diff
@@ -0,0 +1,58 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftCrypto open source project
+//
+// Copyright (c) 2021-2024 Apple Inc. and the SwiftCrypto project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.md for the list of SwiftCrypto project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+import Crypto
+#if canImport(Darwin) || swift(>=5.9.1)
+import Foundation
+#else
+@preconcurrency import Foundation
+#endif
+
+#if canImport(CommonCrypto)
+fileprivate typealias BackingPBKDF2 = CommonCryptoPBKDF2
+#else
+fileprivate typealias BackingPBKDF2 = BoringSSLPBKDF2
+#endif
+
+extension KDF.Insecure {
+    /// An implementation of PBKDF2 key derivation function.
+    public struct PBKDF2 {
+        /// Derives a symmetric key using the PBKDF2 algorithm.
+        ///
+        /// - Parameters:
+        ///    - password: The passphrase, which should be used as a basis for the key. This can be any type that conforms to `DataProtocol`, like `Data` or an array of `UInt8` instances.
+        ///    - salt: The salt to use for key derivation.
+        ///    - hashFunction: The hash function to use for key derivation.
+        ///    - outputByteCount: The length in bytes of resulting symmetric key.
+        ///    - rounds: The number of rounds which should be used to perform key derivation.
+        /// - Returns: The derived symmetric key.
+        public static func deriveKey<Passphrase: DataProtocol, Salt: DataProtocol>(from password: Passphrase, salt: Salt, using hashFunction: HashFunction, outputByteCount: Int, rounds: Int) throws -> SymmetricKey {
```

Comment:
This API generally looks good, but I'm wondering if we should add a minimum round count to avoid folks naively using the number `1`. We should _certainly_ add documentation suggesting a high round count (in the hundreds of thousands), but I can't see any world where we'd want to accept less than 1k.

If we have use-cases that absolutely need lower round counts I'd love to know, but those might be best served the way we serve RSA key sizes: with a second API that is a bit harder to justify typing.

Reply:
I do actually have such a use case. User+password authentication in Oracle typically encrypts the password with 4096 iterations. Hashes it alongside other data using SHA512, does a bit of CBC de- and encryption with it and some more data. **And finally encrypts the end result using another 3 rounds of PBKDF2**. (https://github.com/lovetodream/oracle-nio/blob/7fee78535f68c9f3a451beeed15483bc6ac2877b/Sources/OracleNIO/Messages/Coding/OracleFrontendMessageEncoder.swift#L926-L974)

It might not be the same for every version of Oracle because this is controlled by the db server, but I confirmed it is the default behaviour with Oracle 23ai (newest release).

Reply:
Ace, that's helpful. So I think that this then suggests we want two APIs, one like the one here and one closer that maybe uses the label `unsafeUncheckedRounds` (or similar). 

My desire is to _enable_ the use-case you have, while strongly discouraging users who pick up PBKDF2 without much thought from putting in extraordinarily low values.

Reply:
Sounds good to me 👍

Reply:
I added the `unsafeUncheckedRounds` API method and a `guard` check in the regular API method to throw an incorrect parameter size error whenever the rounds is less than 1000. Please verify if this is acceptable. 

If we now have the unsafe API method, is the 1000 rounds requirement enough though, as it was proposed way back in original RFC with intent to increase number of rounds over time?

Reply:
Probably not. OWASP suggests a minimum round count of 210k for SHA512-based PBKDF2, and so maybe we should consider that the lower bound.